extensions/python/types/PyProcessSession.cpp

```diff
@@ -0,0 +1,265 @@
+#include "types/PyInputStream.h"
+#include "range/v3/algorithm/remove_if.hpp"
+
+namespace org::apache::nifi::minifi::extensions::python {
+
+namespace core = org::apache::nifi::minifi::core;
+
+PyProcessSession::PyProcessSession(std::shared_ptr<core::ProcessSession> session)
+    : session_(std::move(session)) {
+}
+
+std::shared_ptr<core::FlowFile> PyProcessSession::get() {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+
+  auto flow_file = session_->get();
+
+  if (flow_file == nullptr) {
+    return nullptr;
+  }
+
+  flow_files_.push_back(flow_file);
+
+  return flow_file;
+}
+
+void PyProcessSession::transfer(const std::shared_ptr<core::FlowFile>& flow_file,
+                                const core::Relationship& relationship) {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+
+  if (!flow_file) {
+    throw std::runtime_error("Access of FlowFile after it has been released");
+  }
+
+  session_->transfer(flow_file, relationship);
+}
+
+void PyProcessSession::read(const std::shared_ptr<core::FlowFile>& flow_file, BorrowedObject input_stream_callback) {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+
+  if (!flow_file) {
+    throw std::runtime_error("Access of FlowFile after it has been released");
+  }
+
+  session_->read(flow_file, [&input_stream_callback](const std::shared_ptr<io::InputStream>& input_stream) -> int64_t {
+    return Long(Callable(input_stream_callback.getAttribute("process"))(std::weak_ptr(input_stream))).asInt64();
+  });
+}
+
+void PyProcessSession::write(const std::shared_ptr<core::FlowFile>& flow_file, BorrowedObject output_stream_callback) {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+
+  if (!flow_file) {
+    throw std::runtime_error("Access of FlowFile after it has been released");
+  }
+
+  session_->write(flow_file, [&output_stream_callback](const std::shared_ptr<io::OutputStream>& output_stream) -> int64_t {
+    return Long(Callable(output_stream_callback.getAttribute("process"))(std::weak_ptr(output_stream))).asInt64();
+  });
+}
+
+std::shared_ptr<core::FlowFile> PyProcessSession::create(const std::shared_ptr<core::FlowFile>& flow_file) {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+
+  auto result = session_->create(flow_file);
+
+  flow_files_.push_back(result);
+  return result;
+}
+
+void PyProcessSession::remove(const std::shared_ptr<core::FlowFile>& flow_file) {
+  if (!session_) {
+    throw std::runtime_error("Access of ProcessSession after it has been released");
+  }
+  std::shared_ptr<core::FlowFile> result;
+
+  session_->remove(flow_file);
+  ranges::remove_if(flow_files_, [&flow_file](const auto& ff)-> bool { return ff == flow_file; });
```

Comment:
I think we also need `erase` here to actually remove these owning pointers.

Reply:
good catch, fixed it in https://github.com/apache/nifi-minifi-cpp/pull/1504/commits/71a33853c6b5b71e668be9587b052f016388fdc0#diff-c6e6761718cb79a8c101319d280e709c261cf4231184df8fe7c88e68fc432e6dR111